Tribler/community/tunnel/tunnel_community.py

```diff
@@ -739,53 +750,92 @@ def check_create(self, messages):
-                    yield DropMessage(message, "invalid extend request circuit_id")
-                    continue
+            request = self.request_cache.get(u"anon-created", message.payload.circuit_id)
+            if not request:
+                yield DropMessage(message, "invalid extend request circuit_id")
+                continue
+
+            if not message.payload.node_public_key:
+                yield DropMessage(message, "no node public key specified")
+                continue
+
+            if not (message.payload.node_addr or message.payload.node_public_key in request.candidates):
+                yield DropMessage(message, "node public key not in request candidates and no ip specified")
+                continue
+
             yield message
 
     def check_created(self, messages):
         for message in messages:
-            if not self.is_relay(message.payload.circuit_id) and message.payload.circuit_id in self.circuits:
-                request = self.request_cache.get(u"anon-circuit", message.payload.circuit_id)
-                if not request:
-                    yield DropMessage(message, "invalid created response circuit_id")
-                    continue
+            circuit_id = message.payload.circuit_id
+            request = self.request_cache.get(u"anon-circuit", circuit_id)
+            if not request:
+                yield DropMessage(message, "invalid created response circuit_id")
+                continue
+
             yield message
 
     def check_extended(self, messages):
         for message in messages:
-            if not self.is_relay(message.payload.circuit_id):
-                request = self.request_cache.get(u"anon-circuit", message.payload.circuit_id)
-                if not request:
-                    yield DropMessage(message, "invalid extended response circuit_id")
-                    continue
+            request = self.request_cache.get(u"anon-circuit", message.payload.circuit_id)
+            if not request:
+                yield DropMessage(message, "invalid extended response circuit_id")
+                continue
             yield message
 
     def check_pong(self, messages):
         for message in messages:
-            if not self.is_relay(message.payload.circuit_id):
-                request = self.request_cache.get(u"ping", message.payload.identifier)
-                if not request:
-                    yield DropMessage(message, "invalid ping identifier")
+            request = self.request_cache.get(u"ping", message.payload.identifier)
+            if not request:
+                yield DropMessage(message, "invalid ping identifier")
+                continue
+            yield message
+
+    def check_destroy(self, messages):
+        for message in messages:
+            circuit_id = message.payload.circuit_id
+            cand_sock_addr = message.candidate.sock_addr
+
+            if circuit_id in self.relay_from_to:
+                pass
```

Comment:
I copy/pasted that logic from the original on_destroy method. I was too afraid to touch it, so I simply converted into a pass :smile: